File: arrayandstring/kadanealgorithms/_978_LongestTurbulentSubarray.py
def maxTurbulenceSize(arr):
    """
    :type arr: List[int]
    :rtype: int
    """
    l, r = 0, 1
    res, sign = 1, ""  # result list is non-empty and sign is declared as ""

    while r < len(arr):
        if arr[r-1] > arr[r] and sign != ">":
            res = max(res, r-l+1)
            r += 1
            sign = ">"

        elif arr[r-1] < arr[r] and sign != "<":
            res = max(res, r-l+1)
            r += 1
            sign = "<"

        else:
            r = r + 1 if arr[r] == arr[r-1] else r
            l = r - 1
            sign = ""
    return res

def main():
    arr = [9,4,2,10,7,8,8,1,9]
    print(maxTurbulenceSize(arr))

if __name__ == '__main__':
    main()

File: arrayandstring/kadanealgorithms/_978_LongestTurbulentSubarray.java
package arrayandstring.kadanealgorithms;

public class _978_LongestTurbulentSubarray {
    public static int maxTurbulenceSize(int[] arr) {
        int l = 0;
        int r = 1;
        int res = 1;
        String sign = "";
        while (r < arr.length) {
            if (arr[r-1] > arr[r] && !sign.equals(">")) {
                res = Math.max(res, r - l + 1);
                r++;
                sign = ">";
            } else if (arr[r-1] < arr[r] && !sign.equals("<")) {
                res = Math.max(res, r - l + 1);
                r++;
                sign = "<";
            } else {
                r = (arr[r-1] == arr[r])? r + 1: r;
                l = r-1;
                sign = "";
            }
        }

        return res;

    }
    public static void main(String[] args) {
        int[] arr = {9,4,2,10,7,8,8,1,9};
        System.out.println(maxTurbulenceSize(arr));
    }
}
